turbopack/crates/turbopack-core/src/resolve/mod.rs

```diff
@@ -190,24 +197,37 @@ impl ModuleResolveResult {
         })
     }
 
-    pub fn add_affecting_source_ref(&mut self, source: ResolvedVc<Box<dyn Source>>) {
-        self.affecting_sources.push(source);
-    }
-
     pub fn affecting_sources_iter(&self) -> impl Iterator<Item = ResolvedVc<Box<dyn Source>>> + '_ {
         self.affecting_sources.iter().copied()
     }
 
-    fn clone_with_affecting_sources(
-        &self,
-        affecting_sources: Vec<ResolvedVc<Box<dyn Source>>>,
-    ) -> ModuleResolveResult {
+    pub fn is_unresolvable_ref(&self) -> bool {
+        self.primary.is_empty()
+    }
+}
+
+pub struct ModuleResolveResultBuilder {
+    pub primary: FxIndexMap<RequestKey, ModuleResolveResultItem>,
+    pub affecting_sources: Vec<ResolvedVc<Box<dyn Source>>>,
+}
+
+impl From<ModuleResolveResultBuilder> for ModuleResolveResult {
+    fn from(v: ModuleResolveResultBuilder) -> Self {
         ModuleResolveResult {
-            primary: self.primary.clone(),
-            affecting_sources,
+            primary: v.primary.into_iter().collect::<Vec<_>>().into_boxed_slice(),
```

Comment:
```suggestion
            primary: v.primary.into_iter().collect()
```

This would work too?